analysis/syntactic/structinit/structinit.go

```diff
@@ -0,0 +1,552 @@
+				if isZeroAlloc(instr) {
+					return true
+				}
+			case *ssa.MakeInterface:
+				// TODO confirm:
+				// struct converted to an interface will either have been
+				// explicitly allocated previously or is initialized to the zero
+				// value in the instruction itself
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+// isZeroAlloc returns false if there are any writes to any field or sub-field
+// of the struct with types s allocated in alloc in the alloc instruction's
+// block.
+// This means that isZeroAlloc underapproximates zero-allocations because it
+// does not analyze all writes in the program.
+func isZeroAlloc(alloc *ssa.Alloc) bool {
+	fieldAddrs := fieldAddrsOfAlloc(alloc)
+	instrs := alloc.Block().Instrs
+
+	for _, instr := range instrs {
+		if store, ok := instr.(*ssa.Store); ok {
+			if addr, ok := store.Addr.(*ssa.FieldAddr); ok {
+				if _, ok := fieldAddrs[addr]; ok {
+					return false
+				}
+			}
+		}
+	}
+
+	return true
+}
+
+// fieldAddrsOfAlloc returns all the instructions that address a field or
+// sub-field of the struct allocated in alloc.
+func fieldAddrsOfAlloc(alloc *ssa.Alloc) map[*ssa.FieldAddr]struct{} {
+	instrs := alloc.Block().Instrs
+	fieldAddrs := make(map[*ssa.FieldAddr]struct{})
+	vals := map[ssa.Value]struct{}{alloc: {}}
+	for _, instr := range instrs {
+		if addr, ok := instr.(*ssa.FieldAddr); ok {
+			if _, ok := vals[addr.X]; ok {
+				fieldAddrs[addr] = struct{}{}
+				if _, ok := isStructType(addr.Type()); ok {
+					// if the struct field being addressed is a struct,
+					// track all future addresses to it
+					vals[addr] = struct{}{}
+				}
+			}
+		}
+	}
+
+	return fieldAddrs
+}
+
+// structTypes contains both the named struct type
+// (e.g., "[...]syntactic/structinit.structTypes") and its
+// underlying struct type (e.g. "struct { strct: [...] }").
+type structTypes struct {
+	strct *types.Struct
+	named *types.Named
+}
+
+// isStructType returns the named and underlying types of t
+// if it is a struct or pointer to a struct.
+func isStructType(t types.Type) (structTypes, bool) {
+	if t.Underlying() == nil {
+		return structTypes{}, false
+	}
+
+	typ := t
+	if ptr, ok := t.Underlying().(*types.Pointer); ok {
+		typ = ptr.Elem()
+	}
+	if n, ok := typ.(*types.Named); ok {
+		if s, ok := n.Underlying().(*types.Struct); ok {
+			return structTypes{strct: s, named: n}, true
+		}
+	}
+
+	return structTypes{}, false
+}
+
+// alloced is a struct value that was allocated.
+// The value either is the result of an allocation instruction or the struct
+// that was converted to an interface.
+type alloced struct {
+	val   ssa.Value       // val is the allocated value.
+	typs  structTypes     // typs are the types of val.
+	instr ssa.Instruction // instr is the allocation instruction.
+}
+
+// allStructAllocs returns all the instructions in fns that can allocate a value.
+// An allocation instruction is not always explicit:
+// MakeInterface instructions can also "allocate" a value.
```

Comment:
Could changeType also do the same?

Reply:
Good catch! I added some tests and found that the analysis is unsound so I need to fix it